app/models/payment.rb

```diff
@@ -1,5 +1,20 @@
-class Payment < ApplicationRecord
+class PaymentResultValidator < ActiveModel::Validator
+    def validate payment
+        if JSON.parse(payment.payment_result)["order_number"] != Cart.find_by_id(JSON.parse(payment.payment_result)["custom_params"]).order.order_number
+            payment.errors.add :order_number_equality, "Payment's order_number is not matching Order's order_number"
+        end
+    end
+end
+
+
+class Payment < ApplicationRecord 
     has_one :order, dependent: :destroy
+    has_one :refund_reference, dependent: :destroy
+    validates_with PaymentResultValidator, on: :create
+
+    def validate_successful_refund
```

Comment:
Either use this function as validation or scrap it from model for this case and use it as you originally intended

My idea here was that this condition would be used as validation 
```ruby
if @payment.refund_functionality.status == 200
    errors.add(:something)
end
```

And later it will be only called when trying to save object in controller
```ruby
@payment.save # will call this validation
```